File: inst/tinytest/test_focal.R

m <- matrix(0,3,3)
m[c(4,6)] <- c(1,-1)
r <- rast(nrows=3, ncols=3, vals=1:9, crs="+proj=merc")
f <- focal(r, m, na.rm=TRUE)
e <- c(-4, -5, -6, -6, -6, -6, 4, 5, 6)
expect_equal(e, as.vector(values(f)))

f <- focal(r, t(m), na.rm=TRUE)
e <- c(-2, -2, 2, -5, -2, 5, -8, -2, 8)
expect_equal(e, as.vector(values(f)))

#expect_error(focal(r, m, fun=mean, na.rm=TRUE), pattern="[focal]")

#e <- c(-4, -5, -6, -3, -3, -3, 4, 5, 6)
#expect_equal(e, as.vector(values(f)))

m <- matrix(1,3,3)
f <- focal(r, m, na.rm=TRUE)
e <- c(12,21,16,27,45,33,24,39,28)
expect_equal(e, as.vector(values(f)))

f <- focal(r, 3, na.rm=TRUE)
expect_equal(e, as.vector(values(f)))

f <- focal(r, m, na.rm=FALSE, fillvalue=0)
expect_equal(e, as.vector(values(f)))

f <- focal(r, m, na.rm=FALSE)
e <- c(NA, NA, NA, NA, 45, NA, NA, NA, NA)
expect_equal(e, as.vector(values(f)))

f <- focal(r, 3, na.rm=FALSE)
expect_equal(e, as.vector(values(f)))



r <- rast(nrow=3, ncol=3)
values(r) <- 1:ncell(r)
rr <- rast(nrow=3, ncol=3, xmin=0)
values(rr) <- 1:ncell(rr)

f <- focalValues(r)[1,] 
e <- c(NA, NA, NA, 3, 1, 2, 6, 4, 5)
expect_equal(e, f)

f <- focalValues(rr)[1,]
e <- c(NA, NA, NA, NA, 1, 2, NA, 4, 5)
expect_equal(e, f)

f <- as.vector(values(focal(rr, 3, max, na.rm=TRUE)))
e <- c(5, 6, 6, 8, 9, 9, 8, 9, 9)
expect_equal(e, f)

f <- as.vector(values(focal(r, 3, max, na.rm=TRUE)))
e <- c(6, 6, 6, 9, 9, 9, 9, 9, 9)
expect_equal(e, f)

f <- as.vector(values(focal(rr, 3, sum, na.rm=TRUE)))
e <- c(12, 21, 16, 27, 45, 33, 24, 39, 28)
expect_equal(e, f)

f <- as.vector(values(focal(r, 3, sum, na.rm=TRUE)))
e <- c(21, 21, 21, 45, 45, 45, 39, 39, 39)
expect_equal(e, f)

f <- as.vector(values(focal(rr, 3, mean, na.rm=FALSE)))
e <- c(NA, NA, NA, NA, 5, NA, NA, NA, NA)
expect_equal(e, f)

f <- as.vector(values(focal(r, 3, mean, na.rm=FALSE)))
e <- c(NA, NA, NA, 5, 5, 5, NA, NA, NA)
expect_equal(e, f)


r <- rast(ncols=100, nrows=100, ext(0, 10, 0, 10))
values(r) = 1:ncell(r)
r[5,]=NA
f= focal(r, w=5, fun=mean, na.policy="only", na.rm=TRUE, wopt=list(steps=4))
x  = (f - r)
expect_equal(sum(values(x), na.rm=TRUE), 0)

File: python/tests/test_focal.py
"""
Tests ported from inst/tinytest/test_focal.R
"""
import numpy as np
import pytest
import terra as pt
from terra.rast import rast
from terra.values import set_values
from terra.focal import focal


def _vals(r):
    return np.array(r.readValues(0, r.nrow(), 0, r.ncol()), dtype=float)


def test_focal_asymmetric_weight_matrix():
    """Custom weight matrix with +1 and -1."""
    m = np.zeros((3, 3))
    # R: m[c(4,6)] <- c(1,-1)  (col-major: row=1,col=2 → idx3; row=3,col=2 → idx5)
    m[0, 1] = 1
    m[2, 1] = -1
    r = rast(nrows=3, ncols=3)
    r.setCRS("+proj=merc")
    r = set_values(r, list(range(1, 10)))
    f = focal(r, m, na_rm=True)
    expected = [-4, -5, -6, -6, -6, -6, 4, 5, 6]
    np.testing.assert_array_almost_equal(_vals(f), expected)


def test_focal_transposed_weight_matrix():
    """Transposed weight matrix."""
    m = np.zeros((3, 3))
    m[0, 1] = 1
    m[2, 1] = -1
    r = rast(nrows=3, ncols=3)
    r.setCRS("+proj=merc")
    r = set_values(r, list(range(1, 10)))
    f = focal(r, m.T, na_rm=True)
    expected = [-2, -2, 2, -5, -2, 5, -8, -2, 8]
    np.testing.assert_array_almost_equal(_vals(f), expected)


def test_focal_sum_3x3_ones_na_rm():
    """3×3 ones kernel, na_rm=True."""
    m = np.ones((3, 3))
    r = rast(nrows=3, ncols=3)
    r = set_values(r, list(range(1, 10)))
    f = focal(r, m, na_rm=True)
    expected = [12, 21, 16, 27, 45, 33, 24, 39, 28]
    np.testing.assert_array_almost_equal(_vals(f), expected)


def test_focal_sum_integer_window():
    """Integer window size 3 is equivalent to 3×3 ones kernel."""
    r = rast(nrows=3, ncols=3)
    r = set_values(r, list(range(1, 10)))
    expected = [12, 21, 16, 27, 45, 33, 24, 39, 28]
    f = focal(r, 3, na_rm=True)
    np.testing.assert_array_almost_equal(_vals(f), expected)


def test_focal_no_na_fillvalue():
    """na_rm=False but fillvalue=0 gives same result as na_rm=True for interior."""
    m = np.ones((3, 3))
    r = rast(nrows=3, ncols=3)
    r = set_values(r, list(range(1, 10)))
    f = focal(r, m, na_rm=False, fillvalue=0)
    expected = [12, 21, 16, 27, 45, 33, 24, 39, 28]
    np.testing.assert_array_almost_equal(_vals(f), expected)


def test_focal_no_na_edges_become_nan():
    """na_rm=False with default fillvalue=NaN: only the central cell has a full neighbourhood."""
    m = np.ones((3, 3))
    r = rast(nrows=3, ncols=3)
    r = set_values(r, list(range(1, 10)))
    f = focal(r, m, na_rm=False)
    v = _vals(f)
    assert np.isnan(v[0]) and np.isnan(v[1]) and np.isnan(v[2])  # top row
    assert np.isnan(v[3]) and np.isnan(v[5])                      # left/right middle
    assert np.isnan(v[6]) and np.isnan(v[7]) and np.isnan(v[8])  # bottom row
    assert v[4] == pytest.approx(45)


def test_focal_integer_window_no_na_edges():
    m = np.ones((3, 3))
    r = rast(nrows=3, ncols=3)
    r = set_values(r, list(range(1, 10)))
    # focal(r, 3, na_rm=FALSE) should match focal(r, m, na_rm=FALSE)
    f1 = focal(r, m, na_rm=False)
    f2 = focal(r, 3, na_rm=False)
    v1 = _vals(f1)
    v2 = _vals(f2)
    np.testing.assert_array_equal(
        np.isnan(v1), np.isnan(v2),
        err_msg="NaN pattern should match between matrix and integer-window focal"
    )
    mask = ~np.isnan(v1)
    np.testing.assert_array_almost_equal(v1[mask], v2[mask])
